### PR TITLE
update Sentry SDK add features

### PR DESCRIPTION
**Pull request by bruno-garcia**
#### Database Migration
NO

#### Description
Just over 1 year ago I updated the Sentry SDK from 2.x to 3.x. 

* https://github.com/Sonarr/Sonarr/pull/5216/
 
We're about to release 4.0 with new features, performance improvements, etc. So I wanted to get a PR going updating things

This time, besides bumping the SDK, I'm running locally and checking Sentry and I noticed a few things could be improved:

### Mark crashes as 'unhandled'

Throwing an exception that crashes the project currently ends up on `Logger.Fatal` that manually captures an event with Sentry. It ends up like this:

image

I added a Mechanism and marked it as `unhandled` so Sentry shows this as a such:

image

It also helps differentiate in the Issues list:

image

Additionally, in combination with [Release Health](https://docs.sentry.io/product/releases/health/) this should report the _Crash Free Rate_ to get an idea how healthy the service is release over release.
Though it seems we have some issue with this because even with the _unhandled_ flag, it's not reporting as `crashed`: https://github.com/getsentry/sentry-dotnet/issues/2863

### Enable offline caching

In the event the server is offline, events get spooled on disk
By default 30 files max stay on disk, the SDK rotates them

### Set different environments

I noticed the code picks from 4 different DSNs based on the environment it's running.
I didn't change that, but I added code to change the environment it reports in events based on what the RuntimeInfo says.

This way we can filter by `environment` in the UI:

image

### Report Release Status

This is known in Sentry as Release Health as mentioned above. Besides Crash Free rate, it can give you an idea of what versions are still being used in the wild

Here's just my testing but you can imagine this being several releases over a period of time. You can change the time range and environment in this view:

image

### Upload debug files to Sentry

There are a few advantages in doing this:

1. You don't need to ship PDB in the final app, taking less space in the final artifact or server file system
2. The source code can be bundled in the PDB, so Sentry can show you surrounding lines of code

Clicking on the lines:
image

Expand to some additional context:
image

Sentry can also fetch source code and PDBs (to get line numbers) from open source libraries.
New Sentry projects already come configured with the new `NuGet.org` symbol server, but old projects might need to be configured:

New defaults are these. And useful for .NET mainly the `NuGet.org` one.
image

Also added [source link](https://github.com/dotnet/sourcelink/blob/main/docs/README.md), Sentry should show a link "Open on GitHub" to take you to the right file/line (at the right commit sha).
TODO: Not showing up, gotta debug this

#### Todos
- [ ] Tests
- [ ] Wiki Updates

#### Issues Fixed or Closed by this PR

*

### CODE REVIEW

File: src/NzbDrone.Common/Instrumentation/Sentry/SentryTarget.cs
Comment (bruno-garcia) [inline src/NzbDrone.Common/Instrumentation/Sentry/SentryTarget.cs:115]: TODO: Remove before merging.

File: src/NzbDrone.Common/Sonarr.Common.csproj
Comment (bruno-garcia) [inline src/NzbDrone.Common/Sonarr.Common.csproj:14]: We can merge this once `4.0.0` is out but hopefully we get reviews before that so the PR is ready to go.

Also happy to maintain the beta versions as we're close to releasing GA anyway

**Comment by bruno-garcia** [comment]
CI is broken but I have no visibility. Any chance some can share some details?

**Comment by markus101** [comment]
You can log into CI as a guest, the current error is 
```
C:\BuildAgent\work\853acb111afc90d8\src\NzbDrone.Common.Test\InstrumentationTests\SentryTargetFixture.cs(29,28): error CS7036: There is no argument given that corresponds to the required formal parameter 'appFolderInfo' of 'SentryTarget.SentryTarget(string, IAppFolderInfo)'
```

**Comment by bruno-garcia** [comment]
Ah I get that locally, didn't build the whole thing. Thanks @markus101

File: src/NzbDrone.Common/Instrumentation/Sentry/SentryTarget.cs
Comment (bruno-garcia) [inline src/NzbDrone.Common/Instrumentation/Sentry/SentryTarget.cs:115]:
```suggestion
```

**Review by markus101 — APPROVED** [review]
Changes look good, but going to hold off on landing until it's out of beta so we don't push this downstream + we're close to finally getting v4 out of beta.

Thanks for updating again!

File: src/NzbDrone.Common/Sonarr.Common.csproj
Comment (bitsandfoxes) [inline src/NzbDrone.Common/Sonarr.Common.csproj:14]:
```suggestion
    <PackageReference Include="Sentry" Version="4.0.0" />
```

**Comment by bruno-garcia** [comment]
wooo! thanks @bitsandfoxes 

@bitsandfoxes any other features/flags to turn on here for folks or is it ready to go?

File: package.json
Comment (bruno-garcia) [inline package.json:31]:
```suggestion
    "@sentry/browser": "7.100.0",
```

File: src/NzbDrone.Common/Sonarr.Common.csproj
Comment (bruno-garcia) [inline src/NzbDrone.Common/Sonarr.Common.csproj:14]:
```suggestion
    <PackageReference Include="Sentry" Version="4.0.1" />
```

**Comment by bruno-garcia** [comment]
@markus101 bumped Sentry JS and .NET to the latest. Should be good to go! 

Let me know if you need any more changes.

**Comment by bruno-garcia** [comment]
Oh not sure why the build broke 😬

**Comment by bruno-garcia** [comment]
I'm months behind `develop`, let me pull that in and figure it out.

**Comment by bruno-garcia** [comment]
Locally I can build with the build script, except with the --extra-platforms

```
    13>/usr/local/share/dotnet/sdk/6.0.418/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(430,5): error NETSDK1082: There was no runtime pack for Microsoft.NETCore.App available for the specified RuntimeIdentifier 'freebsd-x64'. [/Users/bruno/git/Sonarr/src/NzbDrone.Update/Sonarr.Update.csproj]
```

**Comment by markus101** [comment]
Could the changes in `src/Directory.Build.props`  be causing issues? It shouldn't be a problem, but other builds are working properly.

**Comment by bruno-garcia** [comment]
@markus101 oh isnt' my PR squash merged into `develop`? I'm sure I'll fuck up the rebase lol

**Comment by bruno-garcia** [comment]
build log says:

```
sed: -e expression #1, char 93: unknown option to `s'
```

not related? not much I can make out of the CI log

**Comment by bruno-garcia** [comment]
argh of course

**Comment by bruno-garcia** [comment]
removed the merge commit

**Comment by markus101** [comment]
Looks like this will need a rebase on top of Sonarr/develop still (GitHub is saying it's 160+ commits behind).

**Comment by bruno-garcia** [comment]
> Looks like this will need a rebase on top of Sonarr/develop still (GitHub is saying it's 160+ commits behind).

right, done

**Comment by bruno-garcia** [comment]
```
 sed: -e expression #1, char 93: unknown option to `s'
Error: Process completed with exit code 1.
```

Seems to be it's no longer related to rebasing the PR, right?

**Comment by markus101** [comment]
It took me a bit to dig in and figure it out, but I believe this was caused by the slash in your branch name `ref/sentry-upgrade` that wasn't getting sanitized properly and sed would melt down. I've pushed a change to develop that fixed it in my test build, but please rebase to confirm it's working with external PRs as well.

**Comment by bruno-garcia** [comment]
@markus101 thanks for fixing that! Ci looks good now

**Comment by bruno-garcia** [comment]
Let me know if there are any other issues down the road with Sentry. Feel free to email me at bruno at sentry dot io

**Comment by markus101** [comment]
Thanks Bruno!

**Pull request merged**
